packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,335 @@
+ * the conversion map value.
+ *
+ * @example
+ * For example, for the given string:
+ *
+ * "This is a <span>string</span> with <a>a link</a> and a self-closing
+ * <CustomComponentB/> tag"
+ *
+ * You would have something like this as the conversionMap value:
+ *
+ * ```js
+ * {
+ *     span: <span />,
+ *     a: <a href={ 'https://github.com' } />,
+ *     CustomComponentB: <CustomComponent />,
+ * }
+ * ```
+ *
+ * @param {string}  interpolatedString  The interpolation string to be parsed.
+ * @param {Object}  conversionMap       The map used to convert the string to
+ *                                      a react element.
+ *
+ * @return {Element}  A react element.
+ */
+const createInterpolateElement = ( interpolatedString, conversionMap ) => {
+	indoc = interpolatedString;
+	offset = 0;
+	output = [];
+	stack = [];
+	tokenizer.lastIndex = 0;
+
+	if ( ! isValidConversionMap( conversionMap ) ) {
+		return interpolatedString;
+	}
+
+	do {
+		// twiddle our thumbs
+	} while ( proceed( conversionMap ) );
+
+	return ! output.every( ( element ) => typeof element === 'string' ) ?
+		createElement( Fragment, {}, output ) :
+		output.join( '' );
+};
+
+/**
+ * Validate conversion map.
+ *
+ * A map is considered valid if it's an object and does not have length.
+ *
+ * @private
+ * @param {Object} conversionMap  The map being validated.
+ *
+ * @return {boolean}  True means the map is valid.
+ */
+const isValidConversionMap = ( conversionMap ) => {
+	return typeof conversionMap === 'object' &&
+		typeof conversionMap.length === 'undefined';
+};
+
+/**
+ * This is the iterator over the matches in the string.
+ *
+ * @private
+ * @param {Object} conversionMap The conversion map for the string.
+ *
+ * @return {boolean} true for continuing to iterate, false for finished.
+ */
+function proceed( conversionMap ) {
+	const next = nextToken();
+	const [ tokenType, name, startOffset, tokenLength ] = next;
+	const stackDepth = stack.length;
+	const leadingTextStart = startOffset > offset ? offset : null;
+	if ( ! conversionMap[ name ] ) {
+		if ( stackDepth !== 0 ) {
+			const { stackLeadingText, tokenStart } = stack.pop();
+			output.push( indoc.substr( stackLeadingText, tokenStart ) );
+		}
+		addText();
+		return false;
+	}
+	switch ( tokenType ) {
+		case 'no-more-tokens':
+			if ( stackDepth !== 0 ) {
+				const { stackLeadingText, tokenStart } = stack.pop();
+				output.push( indoc.substr( stackLeadingText, tokenStart ) );
+			}
+			addText();
+			return false;
+
+		case 'self-closed':
+			if ( 0 === stackDepth ) {
+				if ( null !== leadingTextStart ) {
+					output.push(
+						indoc.substr( leadingTextStart, startOffset - leadingTextStart )
+					);
+				}
+				output.push( elementCreator( name, conversionMap[ name ] )() );
+				offset = startOffset + tokenLength;
+				return true;
+			}
+
+			// otherwise we found an inner element
+			addChild(
+				new Component( elementCreator( name, conversionMap[ name ] ), [] ),
```

Comment:
here's where I think we can just push the `conversionMap[ name ]` to the stack instead of `elementCreator( … )` - if we make sure we have all-valid-elements in `validate` then it only leaves the possibility that we have `undefined` for our element here. we can handle that in `Component()`

```js
function Component( component, children = [] ) {
	if ( ! component ) {
		throw new TypeError( conversionMapError( 'found missing component for translation' ) );
	}

 	return {
 		component,
 		children,
 	};
 }
```

Reply:
hmm couple potential issues with that:

- where would we add the key?
- where would we invoke `cloneElement`? 

For the latter, I can see where we'd just invoke `cloneElement` wherever we currently invoke `component.creator( component.children )` but that still leaves the problem of where we get the key unless we enhance `Component` so it does something like this:

```jsx
function Component(name, element, children = []) {
  if (!element) {
    throw new TypeError(conversionMapError('found missing component for translation'));
  }

  return {
    name,
    element,
    children,
  };
}
```

Then we can just do:

```js
output.push( cloneElement( component.element, { key: component.name }, component.children ) ) 
```

Reply:
You seem to have answered your own question in a manner that looks pretty reliable. 😉

Reply:
ha! yup.

Reply:
Implemented in 0c6c0ae.  However, I did not do this in `Component`:

```
if ( ! component ) {
		throw new TypeError( conversionMapError( 'found missing component for translation' ) );
	}
```

Rationale is because the actual translated variation of the incoming string is somewhat out of the control of the client.  If a translation string has a bad token or malformed token because a translator messed up, I think it'd be better to still have the string displayed without crashing the app.  User's will still know something isn't right with the translated string, but the app implementing this function is still usable.

Reply:
That's good reasoning; we should make sure we consider if we could unintentionally expose information we don't mean to by spitting out the raw string.

For example, could we do better if we replaced the broken tags with an empty `<span />`?

I don't know the answer here; just noting that the translation string could end up being confusing out output.

```js
<Icon /> You <ActionLink>did it</ActionLink>!
```
(React preserves the tags in the string as `&lt;` and `&gt;` etc… so this is what we see in the browser)

vs.

```js
You did it!
```
(`<span/>`s omitted because they disappear in the rendered HTML)

That's just one idea. I bet there are other or better fallback mechanisms.

Reply:
Right, I agree there might be better output options available, but then I wonder if it's advantageous to still display the rendered tags in the translation so it's clearer something is wrong with the translation?  Otherwise it may not be obvious and the translation could potentially never get fixed.  I'm on the fence here. 

@akirk, @swissspidy, and @ocean90 - I would appreciate some input from you on this.  To summarize, if there's a unbalanced tag in a translation swapped in for a string (or some other malformed translation in the string breaking original tags), what should we do:

- throw an error (Potentially breaking the app unless there are appropriate error boundaries for handling).
- Leave the malformed tag in place (which will be encoded by react preserving the string) that will get displayed (`&lt;` and `&gt;` etc).  Benefit being the bad translation won't break anything but could appear odd to readers (which hopefully would result in noticing there's something wrong with the translation.